packages/lesswrong/components/form-components/SocialPreviewUpload.tsx

```diff
@@ -0,0 +1,160 @@
+ * the fairly roundabout process that happens on the server to generate the preview description and image. The logic here
+ * is _almost_ exactly the same, but skips some overly complicated or potentially slow steps, namely:
+ * - trimming whitespace
+ * - rendering LaTex
+ * - handling markdown and DraftJS (seeing as these are no longer supported for non admins)
+ *
+ * The complete process:
+ * 1. The document is saved on the server (on submit or "save as draft")
+ *  1.1. originalContents is converted to html using dataToHtml, this handles LaTeX and trimming whitespace also (packages/lesswrong/server/editor/make_editable_callbacks.ts)
+ *  1.2. socialPreviewImageAutoUrl is set to the first image in the document (packages/lesswrong/server/callbacks/postCallbacks.ts)
+ * 2. The document is loaded
+ *  2.1. plaintextDescription is resolved from the html (packages/lesswrong/server/resolvers/revisionResolvers.ts)
+ *  2.2. socialPreviewImageUrl is resolved from the combination of socialPreviewImageId (the image that is (maybe) set explicitly) and socialPreviewAutoUrl (packages/lesswrong/lib/collections/posts/schema.tsx)
+ * 3. The head tags are set in PostsPage (packages/lesswrong/components/posts/PostsPage/PostsPage.tsx)
+ *  3.1 plaintextDescription is truncated
+ *  3.2 socialPreviewImageUrl is just used directly
+ */
+const buildPreviewFromDocument = (document: PostsEditWithLocalData): {description: string | null, fallbackImageUrl: string | null} => {
+  const dataWithDiscardedSuggestions = document.contents?.dataWithDiscardedSuggestions
+  const originalContents = document.contents?.originalContents
+
+  if (!originalContents) return {description: null, fallbackImageUrl: null}
+  if (originalContents.type !== "html" && originalContents.type !== "ckEditorMarkup") {
+    return {
+      description: `<Description preview not supported for this editor type (${originalContents.type}), switch to HTML or EA Forum Docs [Beta] to see the description preview>`,
+      fallbackImageUrl: null,
+    };
+  }
+
+  const html = dataWithDiscardedSuggestions ?? originalContents.data
+  if (!html) return {description: null, fallbackImageUrl: null}
+  
+  const parser = new DOMParser();
+  const htmlDoc = parser.parseFromString(html, "text/html");
+  // get first img tag
+  const img = htmlDoc.querySelector("img");
+
+  const truncatedHtml = truncate(sanitize(html), PLAINTEXT_HTML_TRUNCATION_LENGTH)
+  const plaintextDescription = htmlToText(truncatedHtml, {
+    wordwrap: false,
+    selectors: [
+      { selector: "img", format: "skip" },
+      { selector: "a", options: { ignoreHref: true } },
+    ],
+  }).substring(0, PLAINTEXT_DESCRIPTION_LENGTH);
+  const previewDesc = getPostDescription({...document, contents: {plaintextDescription}})
+  
+  return {description: previewDesc, fallbackImageUrl: img?.getAttribute("src") || null}
+}
+
+const SocialPreviewUpload = ({name, document, updateCurrentValues, clearField, label, croppingAspectRatio, classes}: {
```

Comment:
@Will-Howard Just FYI this component breaks the layout on mobile, prob want to at least wrap the text onto a separate line

Reply:
This is fixed now

Reply:
I still think the text should move onto its own line, since the preview gets really small, plus it still breaks the layout for small screens.

<img width="339" alt="Screen Shot 2023-01-31 at 11 02 46 AM" src="https://user-images.githubusercontent.com/9057804/215815024-add7d785-d804-4d58-ae7f-011126fb1b4f.png">